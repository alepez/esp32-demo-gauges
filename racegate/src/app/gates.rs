#[derive(Debug, serde::Serialize, serde::Deserialize, Default, Clone, Eq, PartialEq)]
pub struct Gate {
    active: bool,
}

#[derive(Debug, serde::Serialize, serde::Deserialize, Default, Clone, Eq, PartialEq)]
pub struct Gates {
    items: [Gate; 16],
}

impl Gates {
    pub fn start_gate(&self) -> &Gate {
        &self.items[0]
    }
    pub fn finish_gate(&self) -> &Gate {
        &self.items[17]
    }
}
