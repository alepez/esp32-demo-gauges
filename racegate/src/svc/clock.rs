#[derive(Default, Debug, Copy, Clone, Eq, PartialEq)]
pub struct LocalInstant(i32);

impl LocalInstant {
    pub fn from_millis(ms: i32) -> Self {
        Self(ms)
    }

    pub fn as_millis(&self) -> i32 {
        self.0
    }
}

pub struct LocalClock {
    start: std::time::Instant,
}

impl Default for LocalClock {
    fn default() -> Self {
        Self {
            start: std::time::Instant::now(),
        }
    }
}

impl LocalClock {
    pub fn now(&self) -> Option<LocalInstant> {
        let t = std::time::Instant::now().checked_duration_since(self.start)?;
        let t_ms = t.as_millis();

        // milliseconds, 32 bits, max 24 days (enough!)
        if t_ms < (i32::MAX as u128) {
            Some(LocalInstant(t_ms as i32))
        } else {
            None
        }
    }
}

#[derive(Default, Debug, Copy, Clone, Eq, PartialEq)]
pub struct CoordinatedInstant(i32);

impl CoordinatedInstant {
    pub fn from_millis(ms: i32) -> Self {
        Self(ms)
    }

    pub fn as_millis(&self) -> i32 {
        self.0
    }
}

pub struct CoordinatedClock<'a> {
    clock: &'a LocalClock,
    offset: LocalOffset,
}

impl<'a> CoordinatedClock<'a> {
    pub fn new(clock: &'a LocalClock, offset: LocalOffset) -> Self {
        Self { clock, offset }
    }

    pub fn now(&self) -> CoordinatedInstant {
        let t = self.clock.now().expect("Cannot get time");
        CoordinatedInstant::from_millis(t.as_millis() + self.offset.as_millis())
    }
}

#[derive(Default, Debug, Copy, Clone, Eq, PartialEq)]
pub struct LocalOffset(i32);

impl LocalOffset {
    pub fn from_millis(ms: i32) -> Self {
        Self(ms)
    }

    pub fn as_millis(&self) -> i32 {
        self.0
    }
}

pub fn calculate_clock_offset(
    coordinator_time: LocalInstant,
    local_time: LocalInstant,
) -> LocalOffset {
    let c = coordinator_time.as_millis();
    let l = local_time.as_millis();
    LocalOffset::from_millis(c - l)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_calculate_clock_offset_when_coordinator_started_before_gate() {
        let coord_time = LocalInstant::from_millis(60_000);
        let local_time = LocalInstant::from_millis(10_000);
        let offset = calculate_clock_offset(coord_time, local_time);
        assert_eq!(offset, LocalOffset::from_millis(50_000));
    }

    #[test]
    fn test_calculate_clock_offset_when_coordinator_started_after_gate() {
        let coord_time = LocalInstant::from_millis(60_000);
        let local_time = LocalInstant::from_millis(110_000);
        let offset = calculate_clock_offset(coord_time, local_time);
        assert_eq!(offset, LocalOffset::from_millis(-50_000));
    }
}
