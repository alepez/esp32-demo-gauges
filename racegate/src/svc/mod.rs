use crate::app::SystemState;
pub use clock::{
    calculate_clock_offset, CoordinatedClock, CoordinatedInstant, LocalClock, LocalInstant,
    LocalOffset,
};
pub use race_node::RaceNode;
pub use std_race_node::StdRaceNode;

mod clock;
pub mod race_node;
mod std_race_node;

pub trait HttpServer {
    fn set_system_state(&self, status: &SystemState);
}
