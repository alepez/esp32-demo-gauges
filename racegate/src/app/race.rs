use std::time::Duration;

use crate::app::gates::Gates;
use crate::svc::CoordinatedInstant;

#[derive(Debug, Default, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct Race {
    start_time: Option<CoordinatedInstant>,
    finish_time: Option<CoordinatedInstant>,
    duration: Option<Duration>,
}

impl Race {
    pub fn set_gates(&mut self, gates: &Gates) {
        let start_time = gates.start_gate().last_activation_time;
        let finish_time = gates.finish_gate().last_activation_time;

        if let Some(start_time) = start_time {
            self.start_time = Some(start_time);
        }

        if let Some(finish_time) = finish_time {
            self.finish_time = Some(finish_time);
        }

        if let Some(start_time) = start_time {
            if let Some(finish_time) = finish_time {
                if start_time > finish_time {
                    self.finish_time = None;
                    self.duration = None;
                } else {
                    let duration_millis = finish_time.as_millis() - start_time.as_millis();
                    self.duration = Some(Duration::from_millis(duration_millis as u64));
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::app::gates::Gate;

    use super::*;

    #[test]
    fn test_race_default() {
        let race = Race::default();
        assert_debug_snapshot!(race);
    }

    #[test]
    fn test_race_with_start_gate_active() {
        let mut race = Race::default();
        race.set_gates(&Gates::new([
            Gate {
                active: true,
                last_activation_time: Some(CoordinatedInstant::from_millis(10_000)),
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
        ]));
        assert_debug_snapshot!(race);
    }

    #[test]
    fn test_race_with_start_and_finish_gates_active() {
        let mut race = Race::default();
        race.set_gates(&Gates::new([
            Gate {
                active: true,
                last_activation_time: Some(CoordinatedInstant::from_millis(10_000)),
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
        ]));
        race.set_gates(&Gates::new([
            Gate {
                active: false,
                last_activation_time: Some(CoordinatedInstant::from_millis(10_000)),
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: true,
                last_activation_time: Some(CoordinatedInstant::from_millis(20_000)),
            },
        ]));
        assert_debug_snapshot!(race);
    }

    #[test]
    fn test_race_with_start_after_finish() {
        let mut race = Race::default();
        race.set_gates(&Gates::new([
            Gate {
                active: true,
                last_activation_time: Some(CoordinatedInstant::from_millis(10_000)),
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
        ]));
        race.set_gates(&Gates::new([
            Gate {
                active: false,
                last_activation_time: Some(CoordinatedInstant::from_millis(10_000)),
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: true,
                last_activation_time: Some(CoordinatedInstant::from_millis(20_000)),
            },
        ]));
        race.set_gates(&Gates::new([
            Gate {
                active: false,
                last_activation_time: Some(CoordinatedInstant::from_millis(30_000)),
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: false,
                last_activation_time: None,
            },
            Gate {
                active: true,
                last_activation_time: Some(CoordinatedInstant::from_millis(20_000)),
            },
        ]));
        assert_debug_snapshot!(race);
    }
}
