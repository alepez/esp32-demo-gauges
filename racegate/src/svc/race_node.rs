use crate::app::gates::Gates;
use crate::hal::gate::GateState;
use crate::svc::CoordinatedInstant;

#[derive(Debug)]
pub enum Error {
    Unknown,
}

pub trait RaceNode {
    fn coordinator_time(&self) -> Option<CoordinatedInstant>;

    fn publish(&self, msg: RaceNodeMessage) -> anyhow::Result<()>;

    fn gates(&self) -> Gates;
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct NodeAddress(u8);

const COORDINATOR_ADDRESS: NodeAddress = NodeAddress(0);
const START_ADDRESS: NodeAddress = NodeAddress(1);
const FINISH_ADDRESS: NodeAddress = NodeAddress(4);

impl NodeAddress {
    pub const fn coordinator() -> Self {
        COORDINATOR_ADDRESS
    }

    pub const fn start() -> Self {
        START_ADDRESS
    }

    pub const fn finish() -> Self {
        FINISH_ADDRESS
    }

    pub const fn is_coordinator(&self) -> bool {
        self.0 == COORDINATOR_ADDRESS.0
    }

    pub const fn is_start(&self) -> bool {
        self.0 == START_ADDRESS.0
    }

    pub const fn is_finish(&self) -> bool {
        self.0 == FINISH_ADDRESS.0
    }

    pub fn as_gate_index(&self) -> Option<usize> {
        if self.0 < 1 {
            None
        } else {
            Some((self.0 as usize) - 1)
        }
    }

    pub const fn unwrap_as_gate_index(&self) -> usize {
        if self.0 < 1 {
            panic!()
        } else {
            (self.0 as usize) - 1
        }
    }
}

pub struct FrameData([u8; RaceNodeMessage::FRAME_SIZE]);

impl FrameData {
    pub fn as_bytes(&self) -> &[u8] {
        self.0.as_slice()
    }
}

impl From<[u8; 16]> for FrameData {
    fn from(value: [u8; 16]) -> Self {
        Self(value)
    }
}

#[derive(Debug, Copy, Clone)]
pub struct GateBeacon {
    pub addr: NodeAddress,
    pub state: GateState,
    pub last_activation_time: Option<CoordinatedInstant>,
}

#[derive(Debug, Copy, Clone)]
pub struct CoordinatorBeacon {
    pub time: CoordinatedInstant,
}

#[derive(Debug)]
pub enum RaceNodeMessage {
    GateBeacon(GateBeacon),
    CoordinatorBeacon(CoordinatorBeacon),
}

impl RaceNodeMessage {
    pub const FRAME_SIZE: usize = 16;

    pub fn data(&self) -> FrameData {
        FrameData::from(self)
    }
}

impl TryFrom<FrameData> for RaceNodeMessage {
    type Error = Error;

    fn try_from(data: FrameData) -> Result<RaceNodeMessage, Error> {
        let msg_id = data.0.first().ok_or(Error::Unknown)?;
        match msg_id {
            1 => Ok(GateBeacon::try_from(data)?.into()),
            2 => Ok(CoordinatorBeacon::try_from(data)?.into()),
            _ => Err(Error::Unknown),
        }
    }
}

impl TryFrom<FrameData> for GateBeacon {
    type Error = Error;

    fn try_from(data: FrameData) -> Result<GateBeacon, Error> {
        let addr = data.0.get(1).ok_or(Error::Unknown)?;
        let addr = NodeAddress(*addr);

        let gate_state = match data.0.get(2) {
            Some(0) => GateState::Inactive,
            Some(1) => GateState::Active,
            _ => GateState::Inactive,
        };

        let last_activation_time = deserialize_u32(&data, 3).ok_or(Error::Unknown)?;

        let last_activation_time = if last_activation_time == 0xFFFFFFFF {
            None
        } else {
            Some(CoordinatedInstant::from_millis(last_activation_time as i32))
        };

        Ok(GateBeacon {
            addr,
            state: gate_state,
            last_activation_time,
        })
    }
}

impl TryFrom<FrameData> for CoordinatorBeacon {
    type Error = Error;

    fn try_from(data: FrameData) -> Result<CoordinatorBeacon, Error> {
        let time = CoordinatedInstant::from_millis(
            deserialize_u32(&data, 1).ok_or(Error::Unknown)? as i32,
        );

        Ok(CoordinatorBeacon { time })
    }
}

impl From<CoordinatorBeacon> for RaceNodeMessage {
    fn from(x: CoordinatorBeacon) -> Self {
        RaceNodeMessage::CoordinatorBeacon(x)
    }
}

impl From<GateBeacon> for RaceNodeMessage {
    fn from(x: GateBeacon) -> Self {
        RaceNodeMessage::GateBeacon(x)
    }
}

fn serialize_system_state(x: &GateBeacon, data: &mut FrameData) {
    data.0[1] = x.addr.0;
    data.0[2] = x.state as u8;

    if let Some(last_activation_time) = x.last_activation_time {
        serialize_u32(last_activation_time.as_millis() as u32, data, 3);
    } else {
        serialize_u32(0xFFFFFFFF, data, 3);
    }
}

fn serialize_coordinator_beacon(x: &CoordinatorBeacon, data: &mut FrameData) {
    serialize_u32(x.time.as_millis() as u32, data, 1);
}

fn serialize_msg_id(msg: &RaceNodeMessage, data: &mut FrameData) {
    let msg_id = match msg {
        RaceNodeMessage::GateBeacon(_) => 1,
        RaceNodeMessage::CoordinatorBeacon(_) => 2,
    };

    data.0[0] = msg_id;
}

fn serialize_u32(x: u32, data: &mut FrameData, offset: usize) {
    data.0[offset] = ((x >> 24) & 0xFF) as u8;
    data.0[offset + 1] = ((x >> 16) & 0xFF) as u8;
    data.0[offset + 2] = ((x >> 8) & 0xFF) as u8;
    data.0[offset + 3] = ((x) & 0xFF) as u8;
}

fn deserialize_u32(data: &FrameData, offset: usize) -> Option<u32> {
    let d0 = *data.0.get(offset)?;
    let d1 = *data.0.get(offset + 1)?;
    let d2 = *data.0.get(offset + 2)?;
    let d3 = *data.0.get(offset + 3)?;
    Some(((d0 as u32) << 24) | ((d1 as u32) << 16) | ((d2 as u32) << 8) | (d3 as u32))
}

impl From<&RaceNodeMessage> for FrameData {
    fn from(msg: &RaceNodeMessage) -> Self {
        let mut data = FrameData::from([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

        serialize_msg_id(msg, &mut data);

        match msg {
            RaceNodeMessage::GateBeacon(x) => serialize_system_state(x, &mut data),
            RaceNodeMessage::CoordinatorBeacon(x) => serialize_coordinator_beacon(x, &mut data),
        };

        data
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::hal::gate::GateState;

    use super::*;

    #[test]
    fn test_serialize_system_state() {
        let x = GateBeacon {
            addr: NodeAddress::start(),
            state: GateState::Active,
            last_activation_time: Some(CoordinatedInstant::from_millis(12345)),
        };

        let msg = RaceNodeMessage::GateBeacon(x);
        let data = msg.data();

        assert_debug_snapshot!(data.as_bytes());
        assert_debug_snapshot!(RaceNodeMessage::try_from(data).unwrap());
    }

    #[test]
    fn test_serialize_coordinator_beacon() {
        let x = CoordinatorBeacon {
            time: CoordinatedInstant::from_millis(2_123_456_789),
        };

        let msg = RaceNodeMessage::CoordinatorBeacon(x);
        let data = msg.data();

        assert_debug_snapshot!(data.as_bytes());
        assert_debug_snapshot!(RaceNodeMessage::try_from(data).unwrap());
    }
}
