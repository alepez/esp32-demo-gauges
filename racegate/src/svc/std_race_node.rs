use std::net::{SocketAddr, UdpSocket};
use std::sync::{Arc, Mutex};
use std::thread::JoinHandle;
use std::time::Duration;

use crate::app::SystemState;
use crate::svc::{RaceNode, RaceNodeMessage};

pub struct StdRaceNode {
    #[allow(dead_code)]
    thread: JoinHandle<()>,
    state: SharedNodeState,
}

struct StdRaceNodeConfig {
    sender_addr: SocketAddr,
    receiver_addr: SocketAddr,
    broadcast_addr: SocketAddr,
}

impl Default for StdRaceNodeConfig {
    fn default() -> Self {
        Self {
            sender_addr: "0.0.0.0:0".parse().unwrap(),
            receiver_addr: "0.0.0.0:6699".parse().unwrap(),
            broadcast_addr: "255.255.255.255:6699".parse().unwrap(),
        }
    }
}

impl StdRaceNode {
    pub fn new() -> anyhow::Result<Self> {
        Self::new_with_config(StdRaceNodeConfig::default())
    }

    fn new_with_config(config: StdRaceNodeConfig) -> anyhow::Result<Self> {
        let StdRaceNodeConfig {
            sender_addr,
            receiver_addr,
            broadcast_addr,
        } = config;

        let state = SharedNodeState::default();

        log::info!("Starting race node");

        let sender = Self::make_sender(sender_addr)?;
        let receiver = Self::make_receiver(receiver_addr)?;

        let thread = Self::spawn_thread(broadcast_addr, state.clone(), sender, receiver);

        Ok(StdRaceNode { thread, state })
    }

    fn spawn_thread(
        broadcast_addr: SocketAddr,
        state: SharedNodeState,
        sender: UdpSocket,
        receiver: UdpSocket,
    ) -> JoinHandle<()> {
        std::thread::Builder::new()
            .stack_size(64 * 1024)
            .spawn(move || loop {
                let msg: Option<RaceNodeMessage> = state.clone().try_into().ok();
                if let Some(msg) = msg {
                    sender.send_to(&msg.data(), broadcast_addr).ok();
                }

                loop {
                    let mut buf = [0u8; RaceNodeMessage::FRAME_SIZE];

                    if let Ok((number_of_bytes, src_addr)) = receiver.recv_from(&mut buf) {
                        if number_of_bytes == RaceNodeMessage::FRAME_SIZE {
                            let msg = RaceNodeMessage::from(buf);
                            let s = SystemState::from(&msg);
                            log::info!("{src_addr} : {:?}", s);
                        }
                    } else {
                        break;
                    }
                }
            })
            .unwrap()
    }

    fn make_receiver(receiver_addr: SocketAddr) -> anyhow::Result<UdpSocket> {
        let receiver = UdpSocket::bind(receiver_addr)?;
        receiver.set_broadcast(true)?;
        receiver.set_read_timeout(Some(Duration::from_millis(40)))?;
        Ok(receiver)
    }

    fn make_sender(sender_addr: SocketAddr) -> anyhow::Result<UdpSocket> {
        let sender = UdpSocket::bind(sender_addr)?;
        sender.set_broadcast(true)?;
        Ok(sender)
    }
}

impl RaceNode for StdRaceNode {
    fn set_system_state(&self, state: &SystemState) {
        self.state
            .0
            .try_lock()
            .as_mut()
            .map(|x| {
                // TODO Change start/finish depending on this node type
                x.this = Some(*state)
            })
            .ok();
    }
}

#[derive(Default)]
struct NodesState {
    start: Option<SystemState>,
    finish: Option<SystemState>,
    this: Option<SystemState>,
}

#[derive(Clone)]
struct SharedNodeState(Arc<Mutex<NodesState>>);

impl Default for SharedNodeState {
    fn default() -> Self {
        SharedNodeState(Arc::new(Mutex::new(NodesState::default())))
    }
}

impl TryFrom<SharedNodeState> for RaceNodeMessage {
    type Error = ();

    fn try_from(x: SharedNodeState) -> Result<Self, Self::Error> {
        x.0.try_lock()
            .ok()
            .and_then(|x| x.this)
            .map(|x| RaceNodeMessage::from(&x))
            .ok_or(())
    }
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use crate::svc::std_race_node::StdRaceNodeConfig;
    use crate::svc::StdRaceNode;

    fn start_config() -> StdRaceNodeConfig {
        StdRaceNodeConfig {
            sender_addr: "127.0.0.1:0".parse().unwrap(),
            receiver_addr: "127.0.0.1:6699".parse().unwrap(),
            broadcast_addr: "127.0.0.255:6699".parse().unwrap(),
        }
    }

    fn finish_config() -> StdRaceNodeConfig {
        StdRaceNodeConfig {
            sender_addr: "127.0.0.2:0".parse().unwrap(),
            receiver_addr: "127.0.0.2:6699".parse().unwrap(),
            broadcast_addr: "127.0.0.255:6699".parse().unwrap(),
        }
    }

    #[test]
    fn test_node_can_be_created_with_default_config() {
        let node = StdRaceNode::new();
        assert!(node.is_ok());
    }

    #[test]
    fn test_two_nodes_can_talk() {
        let start_node = StdRaceNode::new_with_config(start_config());
        let finish_node = StdRaceNode::new_with_config(finish_config());
        std::thread::sleep(Duration::from_secs(5));
    }
}
