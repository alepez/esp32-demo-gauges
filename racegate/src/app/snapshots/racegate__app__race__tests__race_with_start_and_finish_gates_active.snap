---
source: src/app/race.rs
expression: race
---
Race {
    start_time: Some(
        CoordinatedInstant(
            10000,
        ),
    ),
    finish_time: Some(
        CoordinatedInstant(
            20000,
        ),
    ),
    duration: Some(
        10s,
    ),
}
